DuetAI knowledge lookup prompts (#29507)
apache/beam: learning/prompts/documentation-lookup/17_advanced_ai_ml.md

Prompt:
What are AI and ML capabilities in Apache Beam?
Response:

> Apache Beam has a number of built-in [AI and ML capabilities](https://beam.apache.org/documentation/ml/overview/) that enable you to:

Reviewer: Apache Beam has several built-in [AI and ML capabilities](https://beam.apache.org/documentation/ml/overview/) that enable you to:

- Process large datasets for both preprocessing and model inference.
- Conduct exploratory data analysis and smoothly scale up data pipelines in production as part of your MLOps ecosystem.

> - Run your models in production with varying data loads, both in batch and streaming

Reviewer 2: ```suggestion
- Run your models in production with varying data loads, both in batch and streaming pipelines.
```

> See [here](https://beam.apache.org/documentation/patterns/ai-platform/) for common AI Platform integration patterns in Apache Beam.

Reviewer 2: ```suggestion
For common AI platform integration patterns in Apache Beam, see [AI Platform integration patterns](https://beam.apache.org/documentation/patterns/ai-platform/).
```

> See [here](https://beam.apache.org/documentation/patterns/ai-platform/) for common AI Platform integration patterns in Apache Baam.

Author: Done

> The recommended way to implement inference in Apache Beam is by using the [RunInference API](https://beam.apache.org/documentation/sdks/python-machine-learning/). See [here](https://github.com/apache/beam/blob/master/examples/notebooks/beam-ml/run_inference_pytorch_tensorflow_sklearn.ipynb) for more details on how to use RunInference for PyTorch, scikit-learn, and TensorFlow.

Reviewer 2: ```suggestion
The recommended way to implement inference in Apache Beam is by using the [RunInference API](https://beam.apache.org/documentation/sdks/python-machine-learning/). For more information about how to use RunInference for PyTorch, scikit-learn, and TensorFlow, see the [Use RunInference in Apache Beam](https://github.com/apache/beam/blob/master/examples/notebooks/beam-ml/run_inference_pytorch_tensorflow_sklearn.ipynb) example in GitHub.
```

Using pre-trained models in Apache Beam is also supported with [PyTorch](https://github.com/apache/beam/blob/master/examples/notebooks/beam-ml/run_inference_pytorch.ipynb), [Scikit-learn](https://github.com/apache/beam/blob/master/examples/notebooks/beam-ml/run_inference_sklearn.ipynb), and [Tensorflow](https://github.com/apache/beam/blob/master/examples/notebooks/beam-ml/run_inference_tensorflow.ipynb). Running inference on  [custom models](https://beam.apache.org/documentation/ml/about-ml/#use-custom-models) is also supported.

> Apache Beam also supports automatic model refresh, which allows you to update models, hot-swapping them in a running streaming pipeline with no pause in processing the data stream, avoiding downtime. See [here](https://beam.apache.org/documentation/ml/about-ml/#automatic-model-refresh) for more details.

Reviewer 2: ```suggestion
Apache Beam also supports automatically updating the model being used with the `RunInference PTransform` in streaming pipelines without stopping the pipeline. The feature lets you avoid downtime downtime. For more information, see [Automatic model refresh](https://beam.apache.org/documentation/ml/about-ml/#automatic-model-refresh).
```

> Apache Beam also supports automatic model refresh, which allows you to update models, hot swapping them in a running streaming pipeline with no pause in processing the stream of data, avoiding downtime. See [here](https://beam.apache.org/documentation/ml/about-ml/#automatic-model-refresh) for more details.

Author: Done

> More on Apache Beam ML innovations for production can be found [here](https://cloud.google.com/blog/products/ai-machine-learning/dataflow-ml-innovations-on-apache-beam/).

Reviewer 2: ```suggestion
For more information about using machine learning models with Apache Beam, see [Running ML models now easier with new Dataflow ML innovations on Apache Beam](https://cloud.google.com/blog/products/ai-machine-learning/dataflow-ml-innovations-on-apache-beam/).
```

> For more hands-on examples of using Apache Beam ML integration, see [here](https://beam.apache.org/documentation/patterns/bqml/)

Reviewer 2: ```suggestion
For an example that use the Apache Beam ML integration, see [BigQuery ML integration](https://beam.apache.org/documentation/patterns/bqml/).
```

> For more hands-on examples of using Apache Beam ML integration see [here](https://beam.apache.org/documentation/patterns/bqml/)

Reviewer: For more hands-on examples of using Apache Beam ML integration, see [here](https://beam.apache.org/documentation/patterns/bqml/)
